### PR TITLE
add committed index in Progress

### LINKED ISSUE #325: Add a way to figure out followers commit index
Raft doesn't requires follower reports its commit index. But aggregated them can help us find out whether a log is committed in all nodes, which is quite useful in some cases.

For example, in TiKV, when enabling hibernate regions, it's better to let all the followers to commit to latest logs before going to sleep, otherwise follower can wait for a long time before committing and applying logs, which can hurt follower reads.

It can also be used in merge too.

### PR DESCRIPTION
**Pull request by accelsao**
Close #325

### CODE REVIEW

File: src/raft.rs [hunk 1]
```diff
@@ -740,7 +740,15 @@ impl<T: Storage> Raft<T> {
     /// changed (in which case the caller should call `r.bcast_append`).
     pub fn maybe_commit(&mut self) -> bool {
         let mci = self.mut_prs().maximal_committed_index().0;
-        self.raft_log.maybe_commit(mci, self.term)
+        let ok = self.raft_log.maybe_commit(mci, self.term);
+        let self_id = self.id;
+        self.update_committed(self_id);
```
Comment (Fullstop000) [inline src/raft.rs:745]: Why use `self.id` as a `committed_index`?

Comment (accelsao) [reply src/raft.rs:745]: I am trying to pass `self.raft_log.committed` to progress set for updating committed index

**Comment by BusyJay** [comment]
Progress committed index should be the index recorded in followers.

**Comment by accelsao** [comment]
maybe let `Progress committed index` be `HashMap` for recording every raft node's committed index?

**Comment by Fullstop000** [comment]
In the leader's step routine, maybe you can get a follower's `commit_index` from msgs like `MsgAppendResp` or `MsgHeartBeatResp` and update the related follower's `Progress`.

**Comment by accelsao** [comment]
travis build error message with `The command "if [[ $TRAVIS_RUST_VERSION == "stable" && $TRAVIS_OS_NAME == "linux" ]]; then cargo fmt --all -- --check; fi" exited with 1.` 

I have no idea why `cargo fmt` failed.

And, do I need all the commit messages signed?

File: src/raft_log.rs [hunk 2]
```diff
@@ -201,7 +201,7 @@ impl<T: Storage> RaftLog<T> {
         term: u64,
         committed: u64,
         ents: &[Entry],
-    ) -> Option<(u64, u64)> {
+    ) -> Option<(u64, u64, u64)> {
```
Comment (BusyJay) [inline src/raft_log.rs:204]: I don't think it's necessary to be changed.

File: src/raft.rs [hunk 3]
```diff
@@ -1345,6 +1346,9 @@ impl<T: Storage> Raft<T> {
             return;
         }
 
+        // update committed index for the append responder
+        pr.update_committed(m.commit);
```
Comment (BusyJay) [inline src/raft.rs:1350]: It can be moved to L1319.

File: src/progress/mod.rs [hunk 4]
```diff
@@ -179,6 +185,11 @@ impl Progress {
         need_update
     }
 
+    /// update committed_index.
+    pub fn update_committed(&mut self, committed_index: u64) {
+        self.committed_index = committed_index
```
Comment (BusyJay) [inline src/progress/mod.rs:190]: Should skip smaller value.

Comment (accelsao) [reply src/progress/mod.rs:190]: I think the committed index is non-decreasing, so we wont get the smaller value.
What I miss here?

Comment (BusyJay) [reply src/progress/mod.rs]: Message can be reordered.

**Comment by Fullstop000** [comment]
@accelsao DCO is required for every single commit in the PR :( 
One approach is to squash all the commits so far into one commit with your sign-off and force push it

**Comment by accelsao** [comment]
whoops..

**Pull request closed without merging**